controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -289,3 +289,73 @@ func (w *Workload) ForwardEtcdLeadership(ctx context.Context, machine *clusterv1
 	}
 	return nil
 }
+
+// EtcdStatus returns the current status of the etcd cluster
+func (w *Workload) EtcdStatus(ctx context.Context) (EtcdStatus, error) {
+	nodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list control plane nodes")
+	}
+
+	etcdClient, err := w.etcdClientGenerator.forNodes(ctx, nodes.Items)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to create etcd client")
+	}
+
+	members, err := etcdClient.Members(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list etcd members using etcd client")
+	}
+
+	statuses := map[string]EtcdMemberStatus{}
+	for _, member := range members {
+		status := EtcdMemberStatus{Member: member, Responsive: true}
+
+		_, err := etcdClient.EtcdClient.Status(ctx, staticPodName("etcd", member.Name))
+		if err != nil {
+			status.Responsive = false
+			w.Log.Error(err, "error fetching etcd status for member", "member", member.Name)
+		}
+		statuses[member.Name] = status
+	}
+
+	return EtcdStatus{
+		Members: statuses,
+	}, nil
+}
+
+// EtcdStatus is a snapshot of the etcd cluster's status
+type EtcdStatus struct {
+	Members map[string]EtcdMemberStatus
+}
+
+// EtcdMemberStatus is a snapshot of a etcd member's status
+type EtcdMemberStatus struct {
+	*etcd.Member
+	Responsive bool
+}
+
+// FailureTolerance is the amount of members the etcd cluster can afford to
+// lose without losing quorum.
+// Ref: https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#what-is-failure-tolerance
+func (e *EtcdStatus) FailureTolerance(machine *clusterv1.Machine) int {
+	var unresponsive int
+	for _, m := range e.Members {
+		if !m.Responsive {
+			unresponsive++
+		}
+	}
+
+	defaultTolerance := len(e.Members) - (len(e.Members)/2.0 + 1) - unresponsive
+
+	if machine.Status.NodeRef == nil {
+		return defaultTolerance
+	}
+
+	member, found := e.Members[machine.Status.NodeRef.Name]
+	if !found || member.Responsive {
+		return defaultTolerance
+	}
+
+	return defaultTolerance + 1
```

Comment:
Similar here, please add a comment to why we are adding one to the defaultTolerance here vs why we return the defaulttolerance if the member is not present in the map or is a responsive member. It took me a bit to work my way through the logic having previous context, but I think it will be highly confusing to folks who are looking at this without that context.